Add support for normal mapping

Noxime opened the pull request:
Affects #13 

I implemented this by adding a new vertex attribute that's a simple vec3, though while looking up the math I stumbled on kvark's 2011 stackoverflow post about how normals and tangents could be represented more compactly by a quaternion. Thoughts? I thought it was clever, maybe something to consider later on
(https://stackoverflow.com/questions/5255806/how-to-calculate-tangent-and-binormal/5261402#5261402)

Noxime committed:
Add support for normal mapping

I implemented this by adding a new vertex attribute that's a simple
vec3, though while looking up the math I stumbled on kvark's 2011
stackoverflow post about how normals and tangents could be represented
more compactly by a quaternion. Thoughts? I thought it was clever, maybe
something to consider later on
(https://stackoverflow.com/questions/5255806/how-to-calculate-tangent-and-binormal/5261402#5261402)

cwfitzgerald commented on rend3/src/datatypes.rs:
```diff
@@ -226,6 +227,17 @@ impl MeshBuilder {
         self
     }
 
+    /// Add vertex normals to the given mesh.
```
Nit: 
```suggestion
    /// Add vertex tangents to the given mesh.
```

cwfitzgerald commented on rend3/src/datatypes.rs:
```diff
@@ -393,6 +413,72 @@ impl Mesh {
+            let pos2 = positions[idx1 as usize];
+            let pos3 = positions[idx2 as usize];
+
+            let tex1 = uvs[idx0 as usize];
```
If we check the length of the uv buffer too, this also can be unchecked.

cwfitzgerald commented on rend3/src/datatypes.rs:
```diff
@@ -393,6 +413,72 @@ impl Mesh {
+                ((edge1.z * uv2.y) - (edge2.z * uv1.y)) * r,
+            );
+
+            unsafe {
```
There should be a comment saying why this is safe to do.

cwfitzgerald commented on rend3/src/datatypes.rs:
```diff
@@ -393,6 +413,72 @@ impl Mesh {
+        );
+    }
+
+    fn calculate_tangents_for_buffers(
```
My only big concern with this is that there may be case where verts really should be split into multiple (as kvark described). That being said I doubt it will affect most meshes, and we can deal with that in a future PR, doesn't have to be now.

cwfitzgerald reviewed (changes requested):
Thank you so much for tackling this! The code looks very good, just have some smaller comments, then we'll be good to go.

As for quaterions, I totally don't get them, but as long as we can do that as remain compatible with meshes with exported tangents, I'm cool with it.

Could you add support for tangents to the rend3-gltf integration, should be just adding another if for another vertex component.

Noxime committed:
Improve tangent calculation and update gltf example

The example GLTF file didn't include tangents so I added uv coords and
tangents to the file

Noxime commented on examples/gltf/src/main.rs:
```diff
@@ -10,6 +10,12 @@ fn load_gltf(
+        .read_tangents()
+        .unwrap()
+        .map(glam::Vec4::from)
+        .map(From::from)
```
I just realized I've completely forgotten about tangent handedness. In most cases `w` will just be 1, but for example when you use mirrored meshes where UV's go in different directions this can be -1 and means the bitangent should be flipped. Do you think this is important for MVP?

cwfitzgerald replied:
I'm not too experienced with normal mapping, but I'm perfectly fine if you want to hold off on dealing with this in this pr. Just file a follow up issue to track what would have to be fixed.

cwfitzgerald reviewed (changes requested):
One more change: I actually meant the new rend3-gltf crate not the gltf example. Could you add tangents there,then we should be good to go!

cwfitzgerald reviewed (approved):
Don't worry about that last change, I need to make some changes to gltf stuff anyway, so I'll roll them together.

Thank you!
